### PR TITLE
Add documentation for shielding coinbase utxos and payment disclosure

### PR DESCRIPTION
**Pull request by bitcartel**

### CODE REVIEW

File: doc/shield-coinbase.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Shielding Coinbase UTXOs
+
+**Summary**
+
+Use z_shieldcoinbase RPC call to shield coinbase utxos.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: Here "utxos" is lowercase, but it's "UTXOs" elsewhere. Keep capitalization consistent, probably changing to "UTXOs" everywhere is best.

File: doc/shield-coinbase.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+## Background
+
+The current Zcash protocol includes a consensus rule that coinbase rewards must be sent to a shielded address.
+
+## User Experience Challenges
+
+A user can use the z_sendmany RPC call to shield coinbase funds, but the call was not designed for sweeping up many utxos, and offered a suboptimal user experience.
+
+If customers send mining pool payouts to their online wallet, the service provider must sort through UTXOs to to correctly determine the non-coinbase UTXO funds that can be withdrawn or transferred by customers to another transparent address.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: "to" is written twice

File: doc/shield-coinbase.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+## Solution
+
+The z_shieldcoinbase call makes it easy to sweep up coinbase rewards from multiple coinbase utxos across multiple coinbase reward addresses.
+
+    z_shieldcoinbase fromaddress toaddress (fee) (limit)
+
+Where the default fee is 0.0010000 ZEC and the limit is 50.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: "The default fee is 0.0010000 ZEC and the default limit on the maximum number of UTXOs to shield is 50."

File: doc/shield-coinbase.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+The number of coinbase utxos selected for shielding can be adjusted by setting the limit parameter. The default value is 50.
+
+If the limit parameter is set to zero, the zcashd `mempooltxinputlimit` option will be used instead, where the default value for `mempooltxinputlimit` is zero which means no limit.
+
+Any limit is constrained by a hard limit due to the consensus rule defining a maximum transaction size of 100,000 bytes.
+
+In general, the more utxos that are selected, the longer it takes for the transaction to be verified.  Due to the quadratic hashing problem, some miners use the `mempooltxinputlimit` option to reject transactions with a large number of UTXO inputs.
+
+Currently, as of November 2017, there is no commonly agreed upon limit, but as a rule of thumb (a form of emergent consensus) if a transacion has less than 100 UTXO inputs, the transaction will be mined promptly by the majority of mining pools, but if it has say 500 UTXO inputs it might take several days to be mined by a miner who has higher or no limits.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: s/transacion/transaction

Comment (arcalinea) [inline doc/shield-coinbase.md]: "if it has say 500 UTXO inputs" --> "if it has many more UTXO inputs, such as 500"

File: doc/shield-coinbase.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+### Anatomy of a z_shieldcoinbase transaction
+
+The transaction created is a shielded transaction.  It consists of a single joinsplit, which consumes coinbase UTXOs as input, and deposits value at a shielded address, minus any fee.
+
+The number of coinbase UTXOs is determined by a user configured limit.
+
+If no limit is set (set limit parameter and `mempooltxinputlimit` option both to zero) the behaviour of z_shieldcoinbase is consume as many utxos as possible, with `z_shieldcoinbase` constructing a transaction upto the size limit of 100,000 bytes.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: s/is consume/is to consume
s/upto/up to
"(set limit parameter and `mempooltxinputlimit` option both to zero)" --> "(in the case when limit parameter and `mempooltxinputlimit` options are set to zero)

File: doc/shield-coinbase.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+As a result, the maximum numper of inputs that can be selected is
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: s/numper/number
s/is/is:

File: doc/shield-coinbase.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+- P2PKH coinbase utxos ~ 662
+- 2-of-3 multisig P2SH coinbase utxos ~ 244.
+
+Here is an example of using `z_shieldcoinbase` on testnet to shield multi-sig coinbase utxos.
+
+- Block 141042 is almost ~2 MB in size (the maximum size for a block) and contains 1 coinbase reward transaction and 20 transactions, each indivually created by a call to z_shieldcoinbase.
+  - https://explorer.testnet.z.cash/block/0050552a78e97c89f666713c8448d49ad1d7263274422272696187dedf6c0d03
+- Drilling down into a transaction, you can see there is one joinsplit, with 244 inputs (vin) inputs, an no outputs (0 vout).
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: word "inputs" is repeated here: "244 inputs (vin) inputs"
s/an no/and no

File: doc/shield-coinbase.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+The `z_shieldcoinbase` call will lock any selected UTXOs. This prevents the selected UTXOs which are already queued up from being selected for any other send operation.  If the `z_shieldcoinbase` call fails, any locked UTXOs are unlocked.
+
+You can use the RPC call `lockunspent` to see which UTXOs have been locked.  You can also use this call to unlock any UTXOs in the event of an unexpected system failure which leaves UTXOs in a locked state.
+
+### Limits, Performance and Transaction Confirmation
+
+The number of coinbase utxos selected for shielding can be adjusted by setting the limit parameter. The default value is 50.
+
+If the limit parameter is set to zero, the zcashd `mempooltxinputlimit` option will be used instead, where the default value for `mempooltxinputlimit` is zero which means no limit.
```
Comment (arcalinea) [inline doc/shield-coinbase.md]: "is zero which means" --> "is zero, which means"

File: doc/shield-coinbase.md (hunk 1 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 2 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 3 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 8 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 4 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 5 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 6 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 7 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

**Comment by bitcartel** [comment]
@arcalinea Review comments addressed.  Payment disclosure docs moved here too.

File: doc/payment-disclosure.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+### Experimental Feature
+
+This is an experimental feature.  Enable it by launching `zcashd` with flags:
+
+    zcashd -experimentalfeatures -paymentdisclosure -debug=paymentdisclosure -txindex=1
+
+These flags can also be set as options in `zcash.conf`.
+
+All nodes that generate or validate payment disclosures must run with `txindex=1` enabled.
```
Comment (str4d) [inline doc/payment-disclosure.md]: Is this enforced in the code? I don't recall seeing this as an explicit requirement in #2159.

File: doc/payment-disclosure.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+The ZIP describes a method of proving that a payment was sent to a shielded address. In the typical case, this means enabling a sender to present a proof that they transferred funds to a recipient's shielded address. 
+
+[1] https://github.com/zcash/zips/pull/119
+
+### Example Use Case
+
+Alice the customer sends 10 ZEC to Bob the merchant at the shielded address shown on their website.  However, Bob is not sure if he received the funds.
+
+Alice's node is running with payment disclosure enabled, so Alice generates a payment disclosure and provides it to Bob who verifies the payment was made.
```
Comment (str4d) [inline doc/payment-disclosure.md]: `provides it to Bob, who`

File: doc/shield-coinbase.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
+# Shielding Coinbase UTXOs
+
+**Summary**
+
+Use z_shieldcoinbase RPC call to shield coinbase UTXOs.
```
Comment (str4d) [inline doc/shield-coinbase.md]:
```
`z_shieldcoinbase`
```

File: doc/shield-coinbase.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+Miners, Mining pools, Online wallets
+
+## Background
+
+The current Zcash protocol includes a consensus rule that coinbase rewards must be sent to a shielded address.
+
+## User Experience Challenges
+
+A user can use the z_sendmany RPC call to shield coinbase funds, but the call was not designed for sweeping up many utxos, and offered a suboptimal user experience.
```
Comment (str4d) [inline doc/shield-coinbase.md]:
```
`z_sendmany`
```

`UTXOs`

File: doc/shield-coinbase.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+If customers send mining pool payouts to their online wallet, the service provider must sort through UTXOs to correctly determine the non-coinbase UTXO funds that can be withdrawn or transferred by customers to another transparent address.
+
+## Solution
+
+The z_shieldcoinbase call makes it easy to sweep up coinbase rewards from multiple coinbase utxos across multiple coinbase reward addresses.
```
Comment (str4d) [inline doc/shield-coinbase.md]:
```
`z_shieldcoinbase`
```

`UTXOs`

File: doc/shield-coinbase.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+    z_shieldcoinbase fromaddress toaddress (fee) (limit)
+
+Where the default fee is 0.0010000 ZEC and the default limit on the maximum number of UTXOs to shield is 50.
```
Comment (str4d) [inline doc/shield-coinbase.md]: `The default fee`

File: doc/shield-coinbase.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+## Examples
+
+Sweep up coinbase utxos from a transparent address you use for mining:
```
Comment (str4d) [inline doc/shield-coinbase.md]: `UTXOs`

File: doc/shield-coinbase.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+    zcash-cli z_shieldcoinbase tMyMiningAddress zMyPrivateAddress
+
+Sweep up coinbase utxos from multiple transparent addresses to a shielded address:
```
Comment (str4d) [inline doc/shield-coinbase.md]: `UTXOs`

File: doc/shield-coinbase.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+    zcash-cli z_shieldcoinbase "*" zMyPrivateAddress
+
+Sweep up with a fee of 1.23 ZEC:
+
+    zcash-cli z_shieldcoinbase tMyMiningAddress zMyPrivateAddress 1.23
+
+Sweep up with a fee of 0.1 ZEC and set limit on the maximum number of utxos to shield at 25:
```
Comment (str4d) [inline doc/shield-coinbase.md]: `UTXOs` (etc. below)

File: doc/shield-coinbase.md [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+Currently, as of November 2017, there is no commonly agreed upon limit, but as a rule of thumb (a form of emergent consensus) if a transaction has less than 100 UTXO inputs, the transaction will be mined promptly by the majority of mining pools, but if it has many more UTXO inputs, such as 500, it might take several days to be mined by a miner who has higher or no limits.
+
+### Anatomy of a z_shieldcoinbase transaction
+
+The transaction created is a shielded transaction.  It consists of a single joinsplit, which consumes coinbase UTXOs as input, and deposits value at a shielded address, minus any fee.
+
+The number of coinbase UTXOs is determined by a user configured limit.
+
+If no limit is set (in the case when limit parameter and `mempooltxinputlimit` options are set to zero) the behaviour of z_shieldcoinbase is to consume as many utxos as possible, with `z_shieldcoinbase` constructing a transaction up to the size limit of 100,000 bytes.
```
Comment (str4d) [inline doc/shield-coinbase.md]:
```
`z_shieldcoinbase`
```

`UTXOs`

File: doc/payment-disclosure.md (hunk 9 above)
Comment (bitcartel) [reply doc/payment-disclosure.md]: Yes, needs it, but not enforced. I filed a ticket to follow-up https://github.com/zcash/zcash/issues/2758

File: doc/payment-disclosure.md (hunk 10 above)
Comment (bitcartel) [reply doc/payment-disclosure.md]: ok

File: doc/shield-coinbase.md (hunk 11 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok

File: doc/shield-coinbase.md (hunk 12 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ok, hmm, thought i did this already with search and replace.

File: doc/shield-coinbase.md (hunk 13 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ditto

File: doc/shield-coinbase.md (hunk 15 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ditto

File: doc/shield-coinbase.md (hunk 16 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ditto

File: doc/shield-coinbase.md (hunk 17 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ditto

File: doc/shield-coinbase.md (hunk 18 above)
Comment (bitcartel) [reply doc/shield-coinbase.md]: ditto

**Comment by bitcartel** [comment]
@str4d Comments addressed.

**Comment by arcalinea** [comment]
ACK

File: doc/payment-disclosure.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+An optional message can be supplied.  This could be used for a refund address or some other reference, as currently it is not common practice to (ahead of time) include a refund address in the memo field when making a payment.
+
+To validate a payment disclosure, the following RPC call can be used:
+
+    z_validatepaymentdisclosure hexdata
+
+### Example
+
+Generate a payment disclosure for the first joinsplit, second output (index starts from zero):
```
Comment (arcalinea) [inline doc/payment-disclosure.md]: s/joinsplit/JoinSplit ?

Comment (bitcartel) [reply doc/payment-disclosure.md]: On mobile now. Minor. Can fix later.

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**